### PR TITLE
luau 0.504 (new formula)

### PR DESCRIPTION
**Pull request by RayCurse**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/luau.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class Luau < Formula
+  desc "Fast, safe, gradually typed embeddable scripting language derived from Lua"
+  homepage "https://luau-lang.org"
+  url "https://github.com/Roblox/luau/archive/0.504.tar.gz"
+  sha256 "544b3107e1c53f8349b1e327efed55ac29cc20a2983dd23031ab07c5b6944633"
+  license "MIT"
+  revision 0
```
Comment (SMillerDev) [inline Formula/luau.rb:7]: This is the default, no need to add it
```suggestion
```

File: Formula/luau.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  head "https://github.com/Roblox/luau.git", branch: "master"
+  depends_on "cmake" => :build
```
Comment (SMillerDev) [inline Formula/luau.rb:9]: Makes is easier to read
```suggestion
  head "https://github.com/Roblox/luau.git", branch: "master"

  depends_on "cmake" => :build
```

File: Formula/luau.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  def install
+    system "cmake", ".", "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
```
Comment (SMillerDev) [inline Formula/luau.rb:12]:
```suggestion
    system "cmake", ".", *std_cmake_args
```

File: Formula/luau.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    system "cmake", "--build", ".", "--target", "Luau.Repl.CLI", "--config", "RelWithDebInfo"
+    system "cmake", "--build", ".", "--target", "Luau.Analyze.CLI", "--config", "RelWithDebInfo"
```
Comment (SMillerDev) [inline Formula/luau.rb:14]: We don't want to have debug info for everyone
```suggestion
    system "cmake", "--build", ".", "--target", "Luau.Repl.CLI"
    system "cmake", "--build", ".", "--target", "Luau.Analyze.CLI"
```

File: Formula/luau.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    bin.install "luau", "luau-analyze"
```
Comment (SMillerDev) [inline Formula/luau.rb:15]: Can't cmake install this?

File: Formula/luau.rb (hunk 4 above)
Comment (carlocab) [reply Formula/luau.rb:14]: Why don't we want to use the default targets?

File: Formula/luau.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  test do
+    system "echo \"print ('Homebrew is awesome!')\" >> test.lua"
```
Comment (carlocab) [inline Formula/luau.rb:19]:
```suggestion
    (testpath/"test.lua").write "print ('Homebrew is awesome!')\n"
```

File: Formula/luau.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    assert_match "Homebrew is awesome!", shell_output("#{bin}/luau test.lua")
```
Comment (carlocab) [inline Formula/luau.rb:20]: If `luau` reads from stdin you can also use `pipe_output` instead of writing to a file.

File: Formula/luau.rb (hunk 4 above)
Comment (RayCurse) [reply Formula/luau.rb]: If I don't specify the targets, it also builds the tests which is unnecessary.

File: Formula/luau.rb (hunk 5 above)
Comment (RayCurse) [reply Formula/luau.rb:15]: There are no install rules defined in `CMakeLists` file from what I can see. I'm not that familiar with `cmake` so I may be missing something or misunderstanding though.

File: Formula/luau.rb (hunk 7 above)
Comment (RayCurse) [reply Formula/luau.rb:20]: `luau` does read from stdin, but for some reason, using pipe output doesn't work. The command gives the correct output but does not terminate afterwards. I'm not sure how to fix this so I just used a file instead.

File: Formula/luau.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  license "MIT"
+  head "https://github.com/Roblox/luau.git", branch: "master"
+
+  depends_on "cmake" => :build
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "cmake", "--build", ".", "--target", "Luau.Repl.CLI"
+    system "cmake", "--build", ".", "--target", "Luau.Analyze.CLI"
+    bin.install "luau", "luau-analyze"
```
Comment (carlocab) [inline Formula/luau.rb:15]:
```suggestion
    system "cmake", "-S", ".", "-B", "build", *std_cmake_args, "-DLUAU_BUILD_TESTS=OFF"
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
```
Can you try this?

File: Formula/luau.rb (hunk 4 above)
Comment (carlocab) [reply Formula/luau.rb]: You can disable the tests with `LUAU_BUILD_TESTS=OFF`.

File: Formula/luau.rb (hunk 5 above)
Comment (carlocab) [reply Formula/luau.rb:15]: CMake typically generates them for you if you don't specify custom ones.

File: Formula/luau.rb (hunk 8 above)
Comment (RayCurse) [reply Formula/luau.rb:15]: It results in an empty installation error.

Comment (carlocab) [reply Formula/luau.rb:15]: Ok, we can replace the last line with
```ruby
bin.install "build/luau", "build/luau-analyze"
```

File: Formula/luau.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  desc "Fast, safe, gradually typed embeddable scripting language derived from Lua"
+  homepage "https://luau-lang.org"
+  url "https://github.com/Roblox/luau/archive/0.504.tar.gz"
+  sha256 "544b3107e1c53f8349b1e327efed55ac29cc20a2983dd23031ab07c5b6944633"
+  license "MIT"
+  head "https://github.com/Roblox/luau.git", branch: "master"
+
+  depends_on "cmake" => :build
+
+  def install
```
Comment (carlocab) [inline Formula/luau.rb:11]:
```suggestion
  depends_on "cmake" => :build

  on_linux do
    depends_on "gcc"
  end

  fails_with gcc: "5"

  def install
```
This should help with the Linux failure.

**Review by carlocab — APPROVED** [review]
Thanks, @RayCurse.

**Pull request closed without merging**